src/Stoa.ts

```diff
@@ -147,18 +168,124 @@ class Stoa {
+                {
+                    // Save a block just received
+                    await this.ledger_storage.putBlocks(block);
+                    this.pushTestEvent(`save ${height}`);
+                    console.log(`Saved a block with block height of ${height}`);
+
+                    res.status(200).send();
+                }
+                else if (height > expected_height)
+                {
+                    // Recovery is required for blocks that are not received.
+                    // start is `expected_height`
+                    // count is `block_data.header.height - expected_height`
+                    // Record a current block after recovery.
+
+                    this.saving_block = true;
+
+                    let block_height = expected_height;
+                    let max_blocks = height - expected_height;
+
+                    let agora_client = new AgoraClient(this.agora_address, this.agora_port);
+                    let blocks = await agora_client.requestBlocks(block_height, max_blocks);
+
+                    // Save previous block
+                    for (let elem of blocks)
+                    {
+                        let element_height = getBlockHeight(elem);
+                        if (element_height == expected_height)
+                        {
+                            await this.ledger_storage.putBlocks(elem);
+                            this.pushTestEvent(`recover ${element_height}`);
+                            console.log(`Recovered a block with block height of ${element_height}`);
+                            expected_height++;
+                        }
+                        else
+                        {
+                            res.status(500).send("An error occurred while saving");
+                            return;
+                        }
+                    }
+
+                    // Save a block just received
+                    await this.ledger_storage.putBlocks(block);
+                    this.pushTestEvent(`save ${height}`);
+                    console.log(`Saved a block with block height of ${height}`);
+
+                    res.status(200).send();
+                }
+                else
+                {
+                    // Do not save because it is already a saved block.
+
+                    this.pushTestEvent(`pass ${height}`);
+                    console.log(`Ignored a block with block height of ${height}`);
+
+                    res.status(200).send();
+                }
+            }
+            catch (error)
+            {
+                console.error("Failed to save block:\n" + error + "\n" + JSON.stringify(block));
                 res.status(500).send("An error occurred while saving");
-            });
+            }
+            finally
+            {
+                this.saving_block = false;
+            }
         });
     }
+
+    protected pushTestEvent (message: string)
+    {
+
+    }
```

Comment:
A method that does nothing, and is only there for the sake of the derived class, which is there for testing, should be a red flag. This is different from dependency injection because the base method does something.